dali/operators/geometry/transform_base_op.h

```diff
@@ -0,0 +1,189 @@
+      }
+    }
+
+    ndim_ = This().ndim(ws);
+
+    output_descs.resize(1);  // only one output
+    output_descs[0].type = TypeTable::GetTypeInfo(dtype_);
+    output_descs[0].shape = uniform_list_shape(nsamples_, {ndim_, ndim_+1});
+    return true;
+  }
+
+  template <typename T>
+  void RunImplTyped(workspace_t<Backend> &ws, dims<>) {
+    DALI_FAIL(make_string("Unsupported number of dimensions ", ndim_));  // NOLINT
+  }
+
+  template <typename T, int ndim, int... ndims>
+  void RunImplTyped(workspace_t<Backend> &ws, dims<ndim, ndims...>) {
+    if (ndim_ != ndim) {
+      RunImplTyped<T>(ws, dims<ndims...>());
+      return;
+    }
+
+    auto &out = ws.template OutputRef<Backend>(0);
+    out.SetLayout("");  // TODO(janton): Decide what layout we want for transforms
+
+    SmallVector<affine_mat_t<T, ndim>, 128> matrices;
+    matrices.resize(nsamples_);
+    for (int i = 0; i < nsamples_; i++) { 
+      matrices[i] = affine_mat_t<T, ndim>::identity();
+    }
+    This().template DefineTransform<T, ndim>(make_span(matrices), ws);
+    auto out_view = view<T>(out);
+    if (has_input_) {
+      auto &in = ws.template InputRef<Backend>(0);
+      auto in_view = view<T>(in);
+      for (int i = 0; i < nsamples_; i++) {
+        ApplyTransform<T, ndim>(out_view[i].data, in_view[i].data, matrices[i]);
+      }
+    } else {
+      for (int i = 0; i < nsamples_; i++) {
+        ApplyTransform<T, ndim>(out_view[i].data, matrices[i]);
+      }
+    }
+  }
+
+  void RunImpl(workspace_t<Backend> &ws) override {
+    TYPE_SWITCH(dtype_, type2id, T, TRANSFORM_INPUT_TYPES, (
+      using SupportedDims = typename TransformImpl::SupportedDims;
+      RunImplTyped<T>(ws, SupportedDims());
+    ), DALI_FAIL(make_string("Unsupported data type: ", dtype_)));  // NOLINT
+  }
+
+  /**
+   * @brief Infers number of dimension.
+   * Transform implementations are expected to override (shadow) this
+   */
+  int ndim(const workspace_t<Backend> &ws) const {
+    if (has_input_) {
+      return input_transform_ndim(ws);
+    }
+    return 2;  // default
+  }
+
+  int input_transform_ndim(const workspace_t<Backend> &ws) const {
+    assert(has_input_);
+    auto &input = ws.template InputRef<Backend>(0);
+    const auto& shape = input.shape();
+    int ndims = shape[0][0];
+    assert(shape[0][1] == ndims + 1);
+    return ndims;
+  }
+
+ private:
+  template <typename T, int ndim>
+  void ApplyTransform(T *transform_out, const affine_mat_t<T, ndim> &M) {
+    for (int i = 0, k = 0; i < ndim; i++) {
+      for (int j = 0; j < ndim + 1; j++, k++) {
+        transform_out[k] = M(i, j);
+      }
+    }
+  }
+
+  template <typename T, int ndim>
+  void ApplyTransform(T *transform_out, const T* transform_in, const affine_mat_t<T, ndim> &M) {
+    auto mat_in = affine_mat_t<T, ndim>::identity();
+    for (int i = 0, k = 0; i < ndim; i++) {
+      for (int j = 0; j < ndim + 1; j++, k++) {
+        mat_in(i, j) = transform_in[k];
+      }
+    }
```

Comment:
```suggestion
```suggestion
    auto mat_in = *reinterpret_cast<mat<ndim, ndim+1, T>>(transform_in);
```

```

Reply:
we agreed its best to leave it as is